### PR TITLE
add c2d testing

### PR DESCRIPTION
**Pull request by BertKleewein**

### CODE REVIEW

File: python/device/device.py [hunk 1]
```diff
@@ -119,7 +131,16 @@ def __init__(self):
         self.send_message_unacked_failure_count = 200
 
         # How many send_message exceptions do we allow before we fail the test?
-        self.send_message_exception_failure_count = 5
+        self.send_message_exception_failure_count = 1
+
+        # How often do we want the service to send test C2D messages?
+        self.receive_message_interval = 2
```
Comment (drwill-ms) [inline python/device/device.py:137]: in seconds?

File: python/device/device.py [hunk 2]
```diff
@@ -223,6 +265,7 @@ def get_longhaul_metrics(self):
 
         props = {
             "runStartUtc": self.metrics.run_start_utc.isoformat(),
+            "runCurrentTimeUtc": datetime.datetime.utcnow().isoformat(),
```
Comment (drwill-ms) [inline python/device/device.py:268]: start time of the run? latest update time?

File: python/device/device.py [hunk 3]
```diff
@@ -518,9 +580,18 @@ def dispatch_incoming_message_thread(self, worker_thread_info):
                 if thief and thief["pairingId"] == self.pairing_id:
                     cmd = thief["cmd"]
                     if cmd == "pingbackResponse":
+                        print("received {} message with {}".format(cmd, thief["pingbacks"]))
```
Comment (drwill-ms) [inline python/device/device.py:583]: Why print sometimes, and logger.info other times?

File: python/device/device.py [hunk 4]
```diff
@@ -119,7 +131,16 @@ def __init__(self):
...
+        self.send_message_exception_failure_count = 1
+
+        # How often do we want the service to send test C2D messages?
+        self.receive_message_interval = 2
+
+        # How big, in bytes, do we we want the random text size in test C2D messages to be
+        self.receive_message_filler_size = 16 * 1024
+
+        # How many missing C2D messages will cause the test to fail?
+        self.receive_message_missing_message_failure_count = 1
```
Comment (drwill-ms) [inline python/device/device.py:143]: I worry we're being too aggressive about killing a test for very strict conditions. I'd rather we try to go as long as we can (no catastrophic failures), and log the issues we see along the way. Then, from the logs, we can calculate the effective run time based on our (possibly changing) criteria. Even later, if we update our criteria, older runs that we didn't kill, can still be re-evaluated.

Comment (BertKleewein) [reply python/device/device.py:143]: good point.  You've said that before and I think I'm seeing the light.  I'll adjust.

File: python/device/device.py (hunk 3 above)
Comment (BertKleewein) [reply python/device/device.py:583]: oops.  changing to logger.info.

File: python/device/device.py (hunk 2 above)
Comment (BertKleewein) [reply python/device/device.py:268]: renaming

**Review by drwill-ms — APPROVED** [review]

**Pull request merged**